File: 2010_2011_mf_sujet_3_convergences/TPconv_MF.m
%Illustration convergence presque sure
%% Question 1
m=mean(1:6)
%% Question 2
X=runifd(5000,6);
S=cumsum(X);
S2=S./[1:5000]';
%% Question 3
plot(S2)
hold on;
plot(m*ones(1,5000),'r')
%% Question 4
clf
plot(sqrt([1:5000]').*(S2-m))
figure
plot([1:5000]'.*(S2-m))
figure
plot((([1:5000]').^0.1).*(S2-m))
%% Question 5
clf
plotrep(X(1:100))
hold on
plotrep(X(1:500))
plotrep(X(1:1000))
plotrep(X(1:5000))
stairs([1:6],[1:6]/6,'r')
hold off
%% Question 6
clf
X=rand(1,5000);
S=cumsum(X);
S2=S./[1:5000];
plot(S2)
hold on;
plot(0.5*ones(5000,1))
clf
plotrep(X(1:100))
hold on
plotrep(X(1:500))
plotrep(X(1:1000))
plotrep(X(1:5000))
fplot(@(x)x,[0 1])
%% Question 7-8-9-10
clf
U=rand(1,100);
E=(U<1/2);
V=cumsum(E./2.^[1:100]);
plot(V);
hold on
U=rand(1,100);
E=(U<1/2);
V=cumsum(E./2.^[1:100]);
plot(V,'r');
%% Question 11
sigma=sqrt(mean([1:6].^2) - m^2)
clf
for i=1:2000
U=rand(1,30);
E=(U<1/2);
W(i)=sum(E./2.^[1:30]);
end
plotrep(W)
%% Question 13
clf
U=rand(1,30);
E=(U<1/2)+(-1)*(U>1/2);
V=cumsum(E./2.^[1:30]);
plot(V);
hold on
U=rand(1,30);
E=(U<1/2)+(-1)*(U>1/2);
V=cumsum(E./2.^[1:30]);
plot(V,'r');
clf
for i=1:2000
U=rand(1,30);
E=(U<1/2)+(-1)*(U>1/2);
W(i)=sum(E./2.^[1:30]);
end
plotrep(W)

% Theoreme central limite
%% Question 13 
TCLde(10,1000)
TCLde(50,1000)
TCLde(300,1000)
TCLde(1000,1000)
%% Question 14
clf
B=rbinom(5000,10,2/10);
hold on
plotrep(B)
a=[0:10];
poi=exp(-2)*2.^a./factorial(a);
stairs(a,cumsum(poi),'r')
B=rbinom(5000,100,2/100);
plotrep(B)
B=rbinom(5000,1000,2/1000);
plotrep(B)
%%
clf
hold on
stairs([0:10],pbinom([0:10],10,2/10))
stairs([0:10],pbinom([0:10],100,2/100),'r')
stairs([0:10],pbinom([0:10],1000,2/1000),'g')
stairs(a,cumsum(poi),'b')
%%
A=[0 1/3 2/3;1/3 0 2/3;2/3 1/3 0];
P=[0.35 0.25 0.4];
P*A
%%
p0=[0.05 0.9 0.05];
p5=p0*A^5
p10=p0*A^10
p50=p0*A^50
p100=p0*A^100
p0=[0 1 0]
p0*A^10
p0*A^50
clf
for n=1:5000
X=1;
for k=1:1000
X=Markov(X);
end
Y(n)=X;
end
plotrep(Y)
hold on
stairs(cumsum(P),'r')

File: 2010_2011_mf_sujet_3_convergences/TPconv_MF.py
"""Illustration de la convergence (Master MF 2010-2011)."""
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm, binom as binom_dist

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '2010_2011_sujet_3_convergences'))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '2010_2011_sujet_2_simulation_de_lois'))

from plotrep import plotrep
from runifd import runifd
from TCLde import TCLde
from Markov import Markov

# %% Question 1
m = np.mean(np.arange(1, 7))
print(f'm={m}')

# %% Question 2-3
X = runifd(5000, 6)
S2 = np.cumsum(X) / np.arange(1, 5001)
plt.figure()
plt.plot(S2); plt.plot(m * np.ones(5000), 'r')
plt.title('LGN convergence')
plt.show()

# %% Question 4
plt.figure(); plt.plot(np.sqrt(np.arange(1, 5001)) * (S2 - m)); plt.show()
plt.figure(); plt.plot(np.arange(1, 5001) * (S2 - m)); plt.show()
plt.figure(); plt.plot((np.arange(1, 5001)**0.1) * (S2 - m)); plt.show()

# %% Question 5
plt.figure()
for n in [100, 500, 1000, 5000]:
    plotrep(X[:n])
plt.step(np.arange(1, 7), np.arange(1, 7)/6, 'r', where='post')
plt.show()

# %% Question 6
X2 = np.random.rand(5000)
S2b = np.cumsum(X2) / np.arange(1, 5001)
plt.figure(); plt.plot(S2b); plt.plot(0.5 * np.ones(5000)); plt.show()
plt.figure()
for n in [100, 500, 1000, 5000]:
    plotrep(X2[:n])
x01 = np.linspace(0, 1, 200)
plt.plot(x01, x01)
plt.show()

# %% Question 7-8-9-10
plt.figure()
U = np.random.rand(100); E = (U < 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 101)); plt.plot(V)
U = np.random.rand(100); E = (U < 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 101)); plt.plot(V, 'r')
plt.show()

# %% Question 11
sigma = np.sqrt(np.mean(np.arange(1, 7)**2) - m**2)
print(f'sigma={sigma}')
W = np.zeros(2000)
for i in range(2000):
    U = np.random.rand(30); E = (U < 0.5).astype(float)
    W[i] = np.sum(E / 2**np.arange(1, 31))
plt.figure(); plotrep(W); plt.show()

# %% Question 13
plt.figure()
U = np.random.rand(30); E = (U < 0.5).astype(float) + (-1)*(U > 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 31)); plt.plot(V)
U = np.random.rand(30); E = (U < 0.5).astype(float) + (-1)*(U > 0.5).astype(float)
V = np.cumsum(E / 2**np.arange(1, 31)); plt.plot(V, 'r')
plt.show()

W = np.zeros(2000)
for i in range(2000):
    U = np.random.rand(30); E = (U < 0.5).astype(float) + (-1)*(U > 0.5).astype(float)
    W[i] = np.sum(E / 2**np.arange(1, 31))
plt.figure(); plotrep(W); plt.show()

# %% TCL
for n in [10, 50, 300, 1000]:
    plt.figure(); TCLde(n, 1000); plt.title(f'TCL n={n}'); plt.show()

# %% Question 14 — Binomial → Poisson
a_arr = np.arange(9)
poi = np.exp(-2) * 2**a_arr / np.array([np.math.factorial(ai) for ai in a_arr])
plt.figure()
for n, p_val in [(10, 2/10), (100, 2/100), (1000, 2/1000)]:
    B = np.random.binomial(n, p_val, 5000)
    plotrep(B)
plt.step(a_arr, np.cumsum(poi), 'r', where='post')
plt.show()

# CDF comparison
plt.figure()
for n, p_val, col in [(10, 2/10, 'b'), (100, 2/100, 'r'), (1000, 2/1000, 'g')]:
    k = np.arange(11)
    plt.step(k, binom_dist.cdf(k, n, p_val), col, where='post')
plt.step(a_arr, np.cumsum(poi), 'k', where='post', label='Poisson')
plt.legend(); plt.show()

# %% Markov chain
A_mat = np.array([[0, 1/3, 2/3], [1/3, 0, 2/3], [2/3, 1/3, 0]])
P_vec = np.array([0.35, 0.25, 0.4])
print(P_vec @ A_mat)

p0 = np.array([0.05, 0.9, 0.05])
for n in [5, 10, 50, 100]:
    print(f'p{n} =', p0 @ np.linalg.matrix_power(A_mat, n))

Y = np.zeros(5000, dtype=int)
for n in range(5000):
    X = 1
    for k in range(1000):
        X = Markov(X)
    Y[n] = X
plt.figure()
plotrep(Y)
plt.step(np.arange(1, 4), np.cumsum(P_vec), 'r', where='post')
plt.show()
